File: next/electron/main/updateService.ts
import type {
  UpdateCheckResult,
  UpdateDownload,
  UpdateProductId,
  UpdateTarget
} from '../../src/shared/contracts/update'

type UpdateManifest = {
  schemaVersion?: unknown
  products?: unknown
}

type ProductRelease = {
  version: string
  title: string
  notes: string
  prerelease: boolean
  releaseUrl: string
  assets: UpdateAsset[]
}

type UpdateProduct = {
  displayName: string
  releases: ProductRelease[]
}

type UpdateAsset = UpdateDownload & {
  platform: string
  architecture: string
  priority: number
}

export type UpdateClientIdentity = UpdateTarget & {
  productId: UpdateProductId
  packageType?: string
}

type UpdateFetch = (url: string, init: RequestInit) => Promise<Pick<Response, 'ok' | 'status' | 'text'>>

export const currentUpdateProductId: UpdateProductId = 'next-electron'
export const defaultUpdateFeedUrl = 'https://raw.githubusercontent.com/rememberber/MooTool/master/update-manifest.json'
export const defaultReleaseUrl = 'https://github.com/rememberber/MooTool/releases'

export class UpdateService {
  constructor(
    private readonly feedUrl = defaultUpdateFeedUrl,
    private readonly identity: UpdateClientIdentity = {
      productId: currentUpdateProductId,
      platform: process.platform,
      architecture: process.arch
    },
    private readonly fetcher: UpdateFetch = fetch
  ) {}

  async check(currentVersion: string): Promise<UpdateCheckResult> {
    const response = await this.fetcher(this.feedUrl, {
      headers: { Accept: 'application/json', 'User-Agent': 'MooTool-Next' },
      signal: AbortSignal.timeout(10_000)
    })
    if (!response.ok) throw new Error(`Update server returned HTTP ${response.status}`)
    const raw = await response.text()
    if (!raw || raw.length > 2 * 1024 * 1024) throw new Error('Invalid update response')
    const product = parseManifest(raw, this.identity.productId)
    const normalizedCurrent = normalizeVersion(currentVersion)
    const allowPrerelease = parseVersion(normalizedCurrent).prerelease.length > 0
    const eligibleReleases = product.releases.filter((release) => allowPrerelease || !release.prerelease)
    if (eligibleReleases.length === 0) throw new Error(`Update product has no eligible releases: ${this.identity.productId}`)
    const latestRelease = eligibleReleases.reduce((latest, release) => (
      compareVersions(release.version, latest.version) > 0 ? release : latest
    ))
    const normalizedLatest = latestRelease.version
    const available = compareVersions(normalizedLatest, normalizedCurrent) > 0
    return {
      status: available ? 'available' : 'latest',
      productId: this.identity.productId,
      productName: product.displayName,
      currentVersion: normalizedCurrent,
      latestVersion: normalizedLatest,
      releaseUrl: latestRelease.releaseUrl,
      releaseNotes: available ? releaseNotesAfter(eligibleReleases, normalizedCurrent) : '',
      target: {
        platform: this.identity.platform,
        architecture: normalizeArchitecture(this.identity.architecture)
      },
      download: available ? selectUpdateAsset(latestRelease.assets, this.identity) : null,
      checkedAt: new Date().toISOString()
    }
  }
}

export function compareVersions(left: string, right: string): number {
  const a = parseVersion(left)
  const b = parseVersion(right)
  for (let index = 0; index < 3; index += 1) {
    if (a.numbers[index] !== b.numbers[index]) return a.numbers[index] > b.numbers[index] ? 1 : -1
  }
  if (a.prerelease.length === 0 && b.prerelease.length === 0) return 0
  if (a.prerelease.length === 0) return 1
  if (b.prerelease.length === 0) return -1
  for (let index = 0; index < Math.max(a.prerelease.length, b.prerelease.length); index += 1) {
    const leftPart = a.prerelease[index]
    const rightPart = b.prerelease[index]
    if (leftPart === undefined) return -1
    if (rightPart === undefined) return 1
    if (leftPart === rightPart) continue
    const leftNumeric = /^\d+$/.test(leftPart)
    const rightNumeric = /^\d+$/.test(rightPart)
    if (leftNumeric && rightNumeric) return Number(leftPart) > Number(rightPart) ? 1 : -1
    if (leftNumeric !== rightNumeric) return leftNumeric ? -1 : 1
    return leftPart > rightPart ? 1 : -1
  }
  return 0
}

function parseManifest(raw: string, productId: UpdateProductId): UpdateProduct {
  let value: UpdateManifest
  try {
    value = JSON.parse(raw) as UpdateManifest
  } catch {
    throw new Error('Update response is not valid JSON')
  }
  if (value.schemaVersion !== 1 || !isRecord(value.products)) throw new Error('Unsupported update manifest')
  const productValue = value.products[productId]
  if (!isRecord(productValue)) throw new Error(`Update product is not registered: ${productId}`)
  if (productValue.status !== 'active') throw new Error(`Update product is not active: ${productId}`)
  if (typeof productValue.displayName !== 'string' || !Array.isArray(productValue.releases)) {
    throw new Error(`Invalid update product: ${productId}`)
  }
  const releases = productValue.releases.map(parseRelease)
  if (releases.length === 0) throw new Error(`Update product has no releases: ${productId}`)
  if (new Set(releases.map((release) => release.version)).size !== releases.length) {
    throw new Error(`Update product has duplicate release versions: ${productId}`)
  }
  return { displayName: productValue.displayName.slice(0, 120), releases }
}

function parseRelease(value: unknown): ProductRelease {
  if (!isRecord(value) || typeof value.version !== 'string' || typeof value.releaseUrl !== 'string') {
    throw new Error('Invalid release in update manifest')
  }
  const version = normalizeVersion(value.version)
  const prerelease = parseVersion(version).prerelease.length > 0
  if (value.prerelease !== undefined && typeof value.prerelease !== 'boolean') {
    throw new Error(`Invalid prerelease flag for version ${version}`)
  }
  if (typeof value.prerelease === 'boolean' && value.prerelease !== prerelease) {
    throw new Error(`Prerelease flag does not match version ${version}`)
  }
  return {
    version,
    title: typeof value.title === 'string' ? value.title.slice(0, 300) : '',
    notes: typeof value.notes === 'string' ? value.notes.slice(0, 5000) : '',
    prerelease,
    releaseUrl: normalizeHttpsUrl(value.releaseUrl, 'release'),
    assets: Array.isArray(value.assets) ? value.assets.map(parseAsset) : []
  }
}

function parseAsset(value: unknown): UpdateAsset {
  if (!isRecord(value)
    || typeof value.platform !== 'string'
    || typeof value.architecture !== 'string'
    || typeof value.packageType !== 'string'
    || typeof value.fileName !== 'string'
    || typeof value.url !== 'string') {
    throw new Error('Invalid asset in update manifest')
  }
  const fileName = value.fileName.trim()
  if (!fileName || fileName.length > 240 || /[/\\]/.test(fileName)) throw new Error('Invalid update asset file name')
  return {
    platform: value.platform.trim().toLowerCase(),
    architecture: normalizeArchitecture(value.architecture),
    packageType: value.packageType.trim().toLowerCase().slice(0, 40),
    fileName,
    url: normalizeHttpsUrl(value.url, 'asset'),
    priority: typeof value.priority === 'number' && Number.isSafeInteger(value.priority) ? value.priority : 100
  }
}

function selectUpdateAsset(assets: UpdateAsset[], target: UpdateTarget & { packageType?: string }): UpdateDownload | null {
  const platform = target.platform.trim().toLowerCase()
  const architecture = normalizeArchitecture(target.architecture)
  const packageType = target.packageType?.trim().toLowerCase()
  const match = assets
    .filter((asset) => asset.platform === platform && (asset.architecture === architecture || asset.architecture === 'universal'))
    .sort((left, right) => {
      const architectureOrder = Number(left.architecture !== architecture) - Number(right.architecture !== architecture)
      if (architectureOrder !== 0) return architectureOrder
      const packageTypeOrder = Number(Boolean(packageType) && left.packageType !== packageType)
        - Number(Boolean(packageType) && right.packageType !== packageType)
      if (packageTypeOrder !== 0) return packageTypeOrder
      if (left.priority !== right.priority) return left.priority - right.priority
      return packagePreference(platform, left.packageType) - packagePreference(platform, right.packageType)
    })[0]
  return match ? { fileName: match.fileName, packageType: match.packageType, url: match.url } : null
}

function releaseNotesAfter(releases: ProductRelease[], currentVersion: string): string {
  const sections = releases
    .filter((release) => compareVersions(release.version, currentVersion) > 0)
    .sort((left, right) => compareVersions(right.version, left.version))
    .map((release) => {
      const heading = release.title.includes(release.version)
        ? release.title
        : [release.version, release.title].filter(Boolean).join(' — ')
      return [heading, release.notes].filter(Boolean).join('\n')
    })
  const selected: string[] = []
  let length = 0
  for (const section of sections) {
    const addedLength = section.length + (selected.length > 0 ? 2 : 0)
    if (selected.length > 0 && length + addedLength > 12_000) break
    selected.unshift(section.slice(0, 12_000))
    length += addedLength
  }
  return selected.join('\n\n')
}

function packagePreference(platform: string, packageType: string): number {
  const preferences: Record<string, string[]> = {
    darwin: ['zip', 'dmg', 'pkg'],
    win32: ['nsis', 'msi', 'portable', 'zip'],
    linux: ['appimage', 'deb', 'rpm', 'tar.gz']
  }
  const index = (preferences[platform] ?? []).indexOf(packageType)
  return index < 0 ? 100 : index
}

function normalizeArchitecture(value: string): string {
  const normalized = value.trim().toLowerCase()
  if (normalized === 'aarch64') return 'arm64'
  if (normalized === 'amd64') return 'x64'
  return normalized
}

function normalizeHttpsUrl(value: string, kind: string): string {
  let url: URL
  try {
    url = new URL(value)
  } catch {
    throw new Error(`Invalid update ${kind} URL`)
  }
  if (url.protocol !== 'https:') throw new Error(`Update ${kind} URL must use HTTPS`)
  return url.toString()
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value) && typeof value === 'object' && !Array.isArray(value)
}

function normalizeVersion(value: string): string {
  const normalized = value.trim().replace(/^v/i, '')
  parseVersion(normalized)
  return normalized
}

function parseVersion(value: string): { numbers: [number, number, number]; prerelease: string[] } {
  const match = /^(0|[1-9]\d*)(?:\.(0|[1-9]\d*))?(?:\.(0|[1-9]\d*))?(?:-([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?(?:\+[0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*)?$/.exec(value.trim().replace(/^v/i, ''))
  if (!match) throw new Error(`Invalid version: ${value}`)
  const prerelease = match[4]?.split('.') ?? []
  if (prerelease.some((part) => /^0\d+$/.test(part))) throw new Error(`Invalid version: ${value}`)
  return {
    numbers: [Number(match[1]), Number(match[2] ?? 0), Number(match[3] ?? 0)],
    prerelease
  }
}

File: next-flutter/lib/core/update/update_service.dart
import 'dart:convert';
import 'dart:ffi';
import 'dart:io';

import '../../app/product.dart';
import 'package_names.dart';
import 'semver.dart';
import 'update_models.dart';

typedef UpdateTextFetcher = Future<UpdateTextResponse> Function(Uri url);

class UpdateTextResponse {
  const UpdateTextResponse({required this.statusCode, required this.body});
  final int statusCode;
  final String body;
  bool get ok => statusCode >= 200 && statusCode < 300;
}

class UpdateService {
  UpdateService({
    this.feedUrl = defaultUpdateFeedUrl,
    UpdateClientIdentity? identity,
    UpdateTextFetcher? fetcher,
  })  : identity = identity ?? detectUpdateIdentity(),
        fetcher = fetcher ?? defaultUpdateFetcher;

  final String feedUrl;
  final UpdateClientIdentity identity;
  final UpdateTextFetcher fetcher;

  Future<UpdateCheckResult> check(String currentVersion) async {
    if (identity.productId != Product.id) {
      throw StateError('Update productId is compiled to ${Product.id}');
    }
    final uri = Uri.parse(feedUrl);
    if (uri.scheme != 'https') {
      throw FormatException('Update feed must use HTTPS');
    }
    final response = await fetcher(uri);
    if (!response.ok) {
      throw FormatException(
          'Update server returned HTTP ${response.statusCode}');
    }
    if (response.body.isEmpty || response.body.length > 2 * 1024 * 1024) {
      throw FormatException('Invalid update response');
    }
    final normalizedCurrent = normalizeVersion(currentVersion);
    final product = _parseManifest(response.body, identity.productId);
    if (product == null) {
      return _unpublished(normalizedCurrent, Product.displayName);
    }
    final allowPrerelease =
        parseVersion(normalizedCurrent).prerelease.isNotEmpty;
    final eligible = [
      for (final release in product.releases)
        if (allowPrerelease || !release.prerelease) release
    ];
    if (eligible.isEmpty) {
      return _unpublished(normalizedCurrent, product.displayName);
    }
    var latest = eligible.first;
    for (final release in eligible.skip(1)) {
      if (compareVersions(release.version, latest.version) > 0) {
        latest = release;
      }
    }
    final available = compareVersions(latest.version, normalizedCurrent) > 0;
    return UpdateCheckResult(
      status:
          available ? UpdateCheckStatus.available : UpdateCheckStatus.latest,
      productId: Product.id,
      productName: product.displayName,
      currentVersion: normalizedCurrent,
      latestVersion: latest.version,
      releaseUrl: latest.releaseUrl,
      releaseNotes:
          available ? releaseNotesAfter(eligible, normalizedCurrent) : '',
      platform: identity.platform,
      architecture: PackageNames.normalizeArchitecture(identity.architecture),
      checkedAt: DateTime.now().toUtc(),
      download: available ? selectUpdateAsset(latest.assets, identity) : null,
    );
  }

  UpdateCheckResult _unpublished(String current, String name) {
    return UpdateCheckResult(
      status: UpdateCheckStatus.unpublished,
      productId: Product.id,
      productName: name,
      currentVersion: current,
      latestVersion: current,
      releaseUrl: defaultReleaseUrl,
      releaseNotes: '',
      platform: identity.platform,
      architecture: PackageNames.normalizeArchitecture(identity.architecture),
      checkedAt: DateTime.now().toUtc(),
      message: 'unpublished',
    );
  }
}

class _ParsedProduct {
  const _ParsedProduct(this.displayName, this.releases);
  final String displayName;
  final List<ProductRelease> releases;
}

_ParsedProduct? _parseManifest(String raw, String productId) {
  late final Object? decoded;
  try {
    decoded = jsonDecode(raw);
  } on FormatException {
    throw FormatException('Update response is not valid JSON');
  }
  if (decoded is! Map) {
    throw FormatException('Unsupported update manifest');
  }
  final map = Map<String, Object?>.from(decoded);
  if (map['schemaVersion'] != 1 || map['products'] is! Map) {
    throw FormatException('Unsupported update manifest');
  }
  final products = Map<String, Object?>.from(map['products'] as Map);
  final productValue = products[productId];
  if (productValue == null) return null;
  if (productValue is! Map) {
    throw FormatException('Update product is not registered: $productId');
  }
  final product = Map<String, Object?>.from(productValue);
  final status = product['status'] as String? ?? '';
  if (status != 'active') return null;
  if (product['displayName'] is! String || product['releases'] is! List) {
    throw FormatException('Invalid update product: $productId');
  }
  final releases = [
    for (final item in product['releases'] as List) parseRelease(item)
  ];
  if ({for (final release in releases) release.version}.length !=
      releases.length) {
    throw FormatException(
        'Update product has duplicate release versions: $productId');
  }
  return _ParsedProduct(
    _clip(product['displayName'] as String, 120),
    releases,
  );
}

ProductRelease parseRelease(Object? value) {
  if (value is! Map) {
    throw FormatException('Invalid release in update manifest');
  }
  final map = Map<String, Object?>.from(value);
  if (map['version'] is! String || map['releaseUrl'] is! String) {
    throw FormatException('Invalid release in update manifest');
  }
  final version = normalizeVersion(map['version'] as String);
  final prerelease = parseVersion(version).prerelease.isNotEmpty;
  if (map['prerelease'] != null && map['prerelease'] is! bool) {
    throw FormatException('Invalid prerelease flag for version $version');
  }
  if (map['prerelease'] is bool && map['prerelease'] != prerelease) {
    throw FormatException('Prerelease flag does not match version $version');
  }
  return ProductRelease(
    version: version,
    title: map['title'] is String ? _clip(map['title'] as String, 300) : '',
    notes: map['notes'] is String ? _clip(map['notes'] as String, 5000) : '',
    prerelease: prerelease,
    releaseUrl: normalizeHttpsUrl(map['releaseUrl'] as String, 'release'),
    assets: [
      for (final item in map['assets'] as List? ?? const []) parseAsset(item)
    ],
  );
}

UpdateAsset parseAsset(Object? value) {
  if (value is! Map) throw FormatException('Invalid asset in update manifest');
  final map = Map<String, Object?>.from(value);
  if (map['platform'] is! String ||
      map['architecture'] is! String ||
      map['packageType'] is! String ||
      map['fileName'] is! String ||
      map['url'] is! String ||
      map['sha512'] is! String ||
      map['size'] is! num) {
    throw FormatException('Invalid asset in update manifest');
  }
  final fileName = (map['fileName'] as String).trim();
  if (fileName.isEmpty ||
      fileName.length > 240 ||
      fileName.contains('/') ||
      fileName.contains('\\')) {
    throw FormatException('Invalid update asset file name');
  }
  final sha512 = (map['sha512'] as String).trim();
  if (!RegExp(r'^[A-Za-z0-9+/]{86}==$').hasMatch(sha512)) {
    throw FormatException('Invalid update asset SHA-512');
  }
  final size = (map['size'] as num).toInt();
  if (size <= 0) throw FormatException('Invalid update asset size');
  return UpdateAsset(
    platform: (map['platform'] as String).trim().toLowerCase(),
    architecture:
        PackageNames.normalizeArchitecture(map['architecture'] as String),
    packageType: (map['packageType'] as String).trim().toLowerCase(),
    fileName: fileName,
    url: normalizeHttpsUrl(map['url'] as String, 'asset'),
    sha512: sha512,
    size: size,
    priority: map['priority'] is num ? (map['priority'] as num).toInt() : 100,
  );
}

UpdateDownload? selectUpdateAsset(
    List<UpdateAsset> assets, UpdateClientIdentity target) {
  final platform = target.platform.trim().toLowerCase();
  final architecture = PackageNames.normalizeArchitecture(target.architecture);
  final packageType = target.packageType?.trim().toLowerCase();
  final matches = [
    for (final asset in assets)
      if (asset.platform == platform &&
          (asset.architecture == architecture ||
              asset.architecture == 'universal') &&
          (platform != 'darwin' || asset.packageType == 'dmg'))
        asset
  ]..sort((left, right) {
      final architectureOrder = (left.architecture != architecture ? 1 : 0) -
          (right.architecture != architecture ? 1 : 0);
      if (architectureOrder != 0) return architectureOrder;
      if (packageType != null && packageType.isNotEmpty) {
        final packageOrder = (left.packageType != packageType ? 1 : 0) -
            (right.packageType != packageType ? 1 : 0);
        if (packageOrder != 0) return packageOrder;
      }
      if (left.priority != right.priority) {
        return left.priority - right.priority;
      }
      return packagePreference(platform, left.packageType) -
          packagePreference(platform, right.packageType);
    });
  return matches.isEmpty ? null : matches.first.toDownload();
}

int packagePreference(String platform, String packageType) {
  const preferences = {
    'darwin': ['dmg', 'pkg'],
    'win32': ['nsis', 'msi', 'portable', 'zip'],
    'linux': ['appimage', 'deb', 'rpm', 'tar.gz'],
  };
  final index =
      (preferences[platform] ?? const <String>[]).indexOf(packageType);
  return index < 0 ? 100 : index;
}

String releaseNotesAfter(List<ProductRelease> releases, String currentVersion) {
  final newer = [
    for (final release in releases)
      if (compareVersions(release.version, currentVersion) > 0) release
  ]..sort((left, right) => compareVersions(right.version, left.version));
  final selected = <String>[];
  var length = 0;
  for (final release in newer) {
    final heading = release.title.contains(release.version)
        ? release.title
        : [release.version, release.title]
            .where((part) => part.isNotEmpty)
            .join(' — ');
    final section =
        [heading, release.notes].where((part) => part.isNotEmpty).join('\n');
    final clipped =
        section.length > 12000 ? section.substring(0, 12000) : section;
    final added = clipped.length + (selected.isNotEmpty ? 2 : 0);
    if (selected.isNotEmpty && length + added > 12000) break;
    selected.insert(0, clipped);
    length += added;
  }
  return selected.join('\n\n');
}

String normalizeHttpsUrl(String value, String kind) {
  final url = Uri.tryParse(value.trim());
  if (url == null || url.scheme != 'https' || url.host.isEmpty) {
    throw FormatException('Update $kind URL must use HTTPS');
  }
  return url.toString();
}

UpdateClientIdentity detectUpdateIdentity() {
  final platform = Platform.isMacOS
      ? 'darwin'
      : Platform.isWindows
          ? 'win32'
          : 'linux';
  return UpdateClientIdentity(
    platform: platform,
    architecture: currentArchitecture(),
    packageType: detectPackageType(platform),
  );
}

String currentArchitecture() {
  final abi = Abi.current().toString().toLowerCase();
  if (abi.contains('arm64') || abi.contains('aarch64')) return 'arm64';
  if (abi.contains('ia32') || abi.contains('x86')) return 'x86';
  return 'x64';
}

String? detectPackageType(String platform) {
  if (platform == 'darwin') return 'dmg';
  if (platform == 'win32') return 'nsis';
  if (Platform.environment.containsKey('APPIMAGE')) return 'appimage';
  return 'deb';
}

Future<UpdateTextResponse> defaultUpdateFetcher(Uri url) async {
  final client = HttpClient();
  try {
    client.connectionTimeout = const Duration(seconds: 10);
    final request = await client.getUrl(url);
    request.headers.set(HttpHeaders.acceptHeader, 'application/json');
    request.headers.set(HttpHeaders.userAgentHeader, 'MooTool-Next-Flutter');
    final response = await request.close().timeout(const Duration(seconds: 10));
    final body = await utf8.decodeStream(response);
    return UpdateTextResponse(statusCode: response.statusCode, body: body);
  } finally {
    client.close(force: true);
  }
}

String _clip(String value, int max) =>
    value.length <= max ? value : value.substring(0, max);
